packages/repository/src/define-repository-class.ts

```diff
@@ -0,0 +1,192 @@
+
+/**
+ * Signature for repository classes that can be used as the base class for
+ * `define*` functions
+ *
+ * @typeParam M - Model class
+ * @typeParam R - Repository class/interface
+ */
+export interface BaseRepositoryClass<M extends Model, R extends Repository<M>> {
+  /**
+   * The constructor for the generated repository class
+   * @param dataSource - DataSource object
+   */
+  new (modeCtor: Constructor<M>, dataSource: juggler.DataSource): R;
+  prototype: R;
+}
+/**
+ * Create (define) a CRUD repository class for the given model.
+ *
+ * @example
+ *
+ * ```ts
+ * const AddressRepository = defineCrudRepositoryClass(Address);
+ * ```
+ *
+ * @param modelClass - A model class such as `Address`.
+ *
+ * @typeParam M - Model class
+ * @typeParam R - CRUD Repository class/interface
+ */
+export function defineCrudRepositoryClass<
+  M extends Model,
+  R extends CrudRepository<M>
+>(
+  modelClass: typeof Model & {prototype: M},
+  baseRepositoryClass: BaseRepositoryClass<M, R>,
+): CrudRepositoryClass<M, R> {
+  const repoName = modelClass.name + 'Repository';
+  const defineNamedRepo = new Function(
+    'ModelCtor',
+    'BaseRepository',
+    `return class ${repoName} extends BaseRepository {
+      constructor(dataSource) {
+        super(ModelCtor, dataSource);
+      }
+    };`,
+  );
+
+  const repo = defineNamedRepo(modelClass, baseRepositoryClass);
+  assert.equal(repo.name, repoName);
+  return repo;
+}
+
+/**
+ * Create (define) an entity CRUD repository class for the given model.
+ *
+ * @example
+ *
+ * ```ts
+ * const ProductRepository = defineEntityCrudRepositoryClass(Product);
+ * ```
+ *
+ * @param entityClass - An entity class such as `Product`.
+ * @param baseRepositoryClass - Base repository class. Defaults to `DefaultCrudRepository`
+ *
+ * @typeParam E - An entity class
+ * @typeParam IdType - ID type for the entity
+ * @typeParam Relations - Relations for the entity
+ */
+export function defineEntityCrudRepositoryClass<
+  E extends Entity,
+  IdType,
+  Relations extends object = {}
+>(
+  entityClass: typeof Entity & {prototype: E},
+  baseRepositoryClass: BaseRepositoryClass<
+    E,
+    EntityCrudRepository<E, IdType, Relations>
+  > = (DefaultCrudRepository as unknown) as BaseRepositoryClass<
+    E,
+    EntityCrudRepository<E, IdType, Relations>
+  >,
+): EntityCrudRepositoryClass<E, IdType, Relations> {
+  return defineCrudRepositoryClass(entityClass, baseRepositoryClass);
+}
+
+/**
+ * Create (define) a KeyValue repository class for the given entity.
+ *
+ * @example
+ *
+ * ```ts
+ * const ProductKeyValueRepository = defineKeyValueRepositoryClass(Product);
+ * ```
+ *
+ * @param entityClass - An entity class such as `Product`.
+ * @param baseRepositoryClass - Base KeyValue repository class.
+ * Defaults to `DefaultKeyValueRepository`
+ *
+ * @typeParam E - Entity class
+ * @typeParam R - KeyValueRepository class/interface
+ */
+export function defineKeyValueRepositoryClass<
+  E extends Entity,
```

Comment:
Please not that KeyValue repository does not require Entities, it can be used with any model.

https://github.com/strongloop/loopback-next/blob/c016550f51f2f0db74527b96b1ad06d056496448/packages/repository/src/repositories/kv.repository.ts#L25

This was an intentional decision on my side when implementing LB3 key-value APIs, to allow the `key` value to be provided outside of the data stored, thus enabling use cases like storing arbitrary data (including binary blobs) via the key-value API.